Improve "Remove extra paragraph spacing" feature

Palloxin opened the pull request:
repost of #1164 due to conflicts (related code moved to another file)
Currently, it just removes all ``<br>``.
With this change you could:
-remove excessive ``<br>``
-remove some zero-width spaces and ``&nbsp;`` (I found a novel that put them in a p to simulate an extra empty paragraph).
-remove a br after the end of a p. It's just extra space.
-remove a br present right before the end of a p.

EDIT: the feature code has regex chaining now.

Palloxin committed:
Improve "Remove extra paragraph spacing" feature

nyagami commented:
<br> <p class="haha">
failed this test

nyagami commented on android/app/src/main/assets/js/core.js:
```diff
@@ -465,7 +465,10 @@ window.addEventListener('DOMContentLoaded', async () => {
     if (reader.generalSettings.val.removeExtraParagraphSpacing) {
-      html = html.replace(/<\s*br[^>]*>/gi, '\n').replace(/\n{2,}/g, '\n\n');
+      html = html
+        .replace(/([\u200B-\u200D\uFEFF]|&nbsp;)(?<=<p>\1)/g, '')
```
what is these kind of uncode. Do they make sense

Palloxin replied:
They do

nyagami replied:
Could you provide some examples.

nyagami replied:
Why you have to use positive lookbeind and numeric reference instead of simple `<p>&nbsp;</p>`

Palloxin replied:
because ``<`` are more frequent than & and zero width space. The regex will immediately fail without losing time.

nyagami commented on android/app/src/main/assets/js/core.js:
```diff
@@ -465,7 +465,10 @@ window.addEventListener('DOMContentLoaded', async () => {
-      html = html.replace(/<\s*br[^>]*>/gi, '\n').replace(/\n{2,}/g, '\n\n');
+      html = html
+        .replace(/([\u200B-\u200D\uFEFF]|&nbsp;)(?<=<p>\1)/g, '')
+        .replace(/(?:<br>\s*<\/?br>)+/g, '')
```
why adding `+` in the end. Isnt `g` flag is enough

Palloxin replied:
it's faster

nyagami replied:
You should be the one who provide a benchmark :).
But yeah, it maybe true

Palloxin replied:
https://github.com/user-attachments/assets/84dea670-a2b7-4e11-86e6-4da85ad99d16

nyagami commented on android/app/src/main/assets/js/core.js:
```diff
@@ -465,7 +465,10 @@ window.addEventListener('DOMContentLoaded', async () => {
     if (reader.generalSettings.val.removeExtraParagraphSpacing) {
-      html = html.replace(/<\s*br[^>]*>/gi, '\n').replace(/\n{2,}/g, '\n\n');
+      html = html
+        .replace(/([\u200B-\u200D\uFEFF]|&nbsp;)(?<=<p>\1)/g, '')
```
are you sure `(?<=` would working properly
![image](https://github.com/user-attachments/assets/c59ae184-fa7f-43e4-b0bc-d8cdfe6a803e)

Palloxin replied:
With customJS the lookbehind works fine. I am sure if the regex of that feature has different flavour

Palloxin committed:
Adjust in case of p with classes

Palloxin committed:
Lint

Palloxin committed:
Lint - part 2

Palloxin committed:
Lint - part 3

Palloxin committed:
forgot \s*

Palloxin committed:
\u00a exception

found here https://readnovelfull.com/the-wife-of-a-powerful-family-huo-shao-how-dare-you-flirt-with-me/chapter-1.html

Palloxin committed:
Better

Palloxin committed:
Fix

Palloxin committed:
fix lint

Palloxin committed:
forgot ``?:``

Palloxin committed:
Remove zero width

Removed until a concrete source example has them.

Palloxin committed:
fix lint

Palloxin committed:
lint

Palloxin committed:
better v2

Palloxin committed:
better 2.1

Palloxin committed:
Lint

Palloxin committed:
\s

Palloxin committed:
Exception: double br are used to simulate paragraphs

NovelUpdates - The Extra of The Bad Ending Game, chapter 159

Palloxin commented:
Double br exception:
![cF6sm6l4](https://github.com/user-attachments/assets/6f48b824-6be1-405a-a40a-c22b4d97717c)
from
``/(?:<br>\s*<\/?br>\s*)+/``
to
``/(?:<br>\s*<\/?br>\s*)+(?![^<])/``

Palloxin committed:
double br exception correction

Palloxin committed:
-

Palloxin commented:
I think it's ready now

Palloxin committed:
better exception

Palloxin committed:
Lint

Palloxin committed:
flaw correction

doest match ``ahahah<br><br><br><br><br>ahahahaha``

Palloxin committed:
fix little flaw

Palloxin committed:
Shortening regex

Palloxin committed:
faster regex

Palloxin committed:
comment + faster regex

Palloxin committed:
lint

Palloxin committed:
lint

Palloxin committed:
Improve double br exception

Palloxin committed:
Correct flaw

Palloxin committed:
forgot

Palloxin committed:
\b for p

Palloxin committed:
Revert correct flaw

Palloxin committed:
Minimal faster regex

Palloxin committed:
remove useless piece

Palloxin committed:
forgot

Palloxin committed:
Faster regex

Palloxin committed:
very minor change

Palloxin committed:
Max level fast

Palloxin committed:
forgot a thing

Palloxin committed:
missing \s

Palloxin committed:
plus ultra

Palloxin committed:
suppefast

Palloxin committed:
ops

Palloxin committed:
u200b
